### PR TITLE
feat: binary artifact display

### PR DESCRIPTION
**Pull request by kinsolee**
# Changes
1. add structs for binary artifact properties
2. make subResources option being able to used without listOption
3. add project-level artifact interfaces
4. add project artifact http routes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [`spec` PR link](https://github.com/katanomi/spec) included
- [ ] Follows the [commit message standard](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md#commits)
- [ ] Meets the [contributing guidelines](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md) (including
  functionality, content, code)
- [ ] Test cases with documentation and functionality works as expected using current and related github repos (MUST deploy and check)
- [ ] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

### CODE REVIEW

File: apis/artifacts/v1alpha1/artifact_properties.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+const (
+	// BinaryArtifactTypeDirectory for directory type
+	BinaryArtifactTypeDirectory = "directory"
+	// BinaryArtifactTypeFile for file type
+	BinaryArtifactTypeFile = "file"
+)
+
+// ArtifactTypeBinaryProperties properties for binary artifact type
+type ArtifactTypeBinaryProperties struct {
```
Comment (kinsolee) [inline apis/artifacts/v1alpha1/artifact_properties.go:27]: 针对二进制制品仓库类型的参数

File: apis/meta/v1alpha1/artifact_types.go [hunk 2]
```diff
@@ -146,3 +146,14 @@ func ArtifactResourceAttributes(verb string) authv1.ResourceAttributes {
 		Verb:     verb,
 	}
 }
+
+// ArtifactFileSubResourceAttributes returns artifacts/files SubResourceAttribute object to be used in a filter
+func ArtifactFileSubResourceAttributes(verb string) authv1.ResourceAttributes {
```
Comment (kinsolee) [inline apis/meta/v1alpha1/artifact_types.go:151]: 审计支持使用subresource，将制品文件看出是制品的子资源

File: apis/meta/v1alpha1/listmeta_types.go [hunk 3]
```diff
@@ -77,14 +79,17 @@ type ListOptions struct {
 	// +optional
 	Search url.Values `json:",inline"`
 
-	// Subresoures for listing
+	// Sort for listing
+	Sort []SortOptions `json:"sort"`
+}
+
+// SubResourcesOptions options for subresources
```
Comment (kinsolee) [inline apis/meta/v1alpha1/listmeta_types.go:86]: 将 SubResources 作为单独的option，可以独立于ListOption 使用

File: apis/meta/v1alpha1/listmeta_types.go [hunk 4]
```diff
@@ -169,6 +174,16 @@ type ArtifactOptions struct {
...
 }
 
+// ProjectArtifactOptions for project artifact
+type ProjectArtifactOptions struct {
+	// project name
+	Project string `json:"project"`
+	// artifact name
+	Artifact string `json:"artifact"`
+	// subresources
+	SubResourcesOptions `json:",inline"`
```
Comment (kinsolee) [inline apis/meta/v1alpha1/listmeta_types.go:184]: 这里是为了限制obs分配到桶下一级目录的场景，需要做过滤和权限校验，只能访问分配到的subResources(一级目录)。

File: plugin/client/interface.go [hunk 5]
```diff
@@ -133,12 +134,42 @@ type ArtifactLister interface {
...
+	ListProjectArtifacts(ctx context.Context, params metav1alpha1.ArtifactOptions, option metav1alpha1.ListOptions) (*metav1alpha1.ArtifactList, error)
+}
+
 // ArtifactGetter get artifact detail
 type ArtifactGetter interface {
 	Interface
 	GetArtifact(ctx context.Context, params metav1alpha1.ArtifactOptions) (*metav1alpha1.Artifact, error)
 }
 
+// ProjectArtifactGetter get artifact detail
```
Comment (kinsolee) [inline plugin/client/interface.go:149]: 添加新的 Project 级别下的Artifact，不区分哪个 Repository 了，避免跟旧接口混淆。

**Review by airycanon — DISMISSED** [review]

**Review by airycanon — APPROVED** [review]

**Review by airycanon — APPROVED** [review]

**Pull request merged**